### PR TITLE
client connection

### PR DESCRIPTION
**Pull request by StringNick**

### CODE REVIEW

**Comment by StringNick** [comment]
its draw, should I fix it bugs, is it needed for this project?

**Comment by Allenxuxu** [comment]
> its draw, should I fix it bugs, is it needed for this project?

Yes，gev needs it. But It needs better design.
For example, we need to reuse event loop to control multiple client.

**Comment by StringNick** [comment]
U mean, that to reuse event-loop if we need connection pool, or we add option for passing event loop in newclient

**Comment by MrChang0** [comment]
sorry, your design of client is not good. I wander know that why use this client. why not use origin gonet to write client(your code more like it).
and you can learn [muduo client](https://github.com/chenshuo/muduo/blob/master/muduo/net/TcpClient.cc) to learn how tcpclient work in eventloop.
best wishes.

**Comment by StringNick** [comment]
> sorry, your design of client is not good. I wander know that why use this client. why not use origin gonet to write client(your code more like it).
> and you can learn [muduo client](https://github.com/chenshuo/muduo/blob/master/muduo/net/TcpClient.cc) to learn how tcpclient work in eventloop.
> best wishes.

origin gonet in microservices with own load-balancing is heavy loaded of syscalls

**Comment by StringNick** [comment]
pushed version with connector, u can write ur own connection pool, use different clients with differents endpoint servers, i can write later connection pool, w8 comments for fixing

File: client.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
+package gev
+
+import (
+	"errors"
+	"github.com/Allenxuxu/gev/connection"
+	"github.com/Allenxuxu/gev/eventloop"
```
Comment (Allenxuxu) [inline client.go:6]: goimports

File: client.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	"golang.org/x/sys/unix"
+	"net"
+	"runtime"
+	"syscall"
+	"time"
+)
+
+type Connector struct {
+	workLoops   []*eventloop.EventLoop
+	opts        *Options
```
Comment (Allenxuxu) [inline client.go:21]: I think client needs an `options` of his own

File: client.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	timingWheel *timingwheel.TimingWheel
```
Comment (Allenxuxu) [inline client.go:22]: If you don't need it for the time being, you should not add it.

File: client.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+			case <-l:
+				err = errors.New("timeout")
+				_ = unix.Close(fd)
+				return err
+
+			default:
+				wFdSet := &unix.FdSet{}
+				wFdSet.Set(fd)
+
+				n, err = unix.Select(fd+1, wFdSet, wFdSet, nil, &unix.Timeval{Sec: 1, Usec: 0})
```
Comment (Allenxuxu) [inline client.go:93]: use poller?

**Review by Allenxuxu — CHANGES_REQUESTED** [review]

File: client.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+func NewConnector(opts ...Option) (connector *Connector, err error) {
+	connector = new(Connector)
+	connector.opts = newOptions(opts...)
+
+	connector.timingWheel = timingwheel.NewTimingWheel(connector.opts.tick, connector.opts.wheelSize)
+	if connector.opts.NumLoops <= 0 {
+		connector.opts.NumLoops = runtime.NumCPU()
+	}
+
+	wloops := make([]*eventloop.EventLoop, connector.opts.NumLoops)
```
Comment (MrChang0) [inline client.go:159]: client dont need wloops, because one client 'with' one fd. 
btw, you can design like 'ClientPool --->  EventLoop ---> Client'.

**Review by MrChang0 — CHANGES_REQUESTED** [review]

File: client.go (hunk 5 above)
Comment (StringNick) [reply client.go:159]: wloops is for something like clientpool, but its connector, u can use one global object on APP to create multiple connections to different endpoints

File: client.go (hunk 3 above)
Comment (StringNick) [reply client.go:22]: u using timingwheel in connection struct, i pass it in new conn

File: client.go (hunk 5 above)
Comment (MrChang0) [reply client.go:159]: oh I see. this comment for old version. QAQ. 
clientpoll for connector is right.

File: client.go (hunk 4 above)
Comment (StringNick) [reply client.go:94]: i dont think ur poller is ready for that, select got easy timeout

File: client.go (hunk 2 above)
Comment (StringNick) [reply client.go:22]: dunno, should i ? options particularly close same

Comment (Allenxuxu) [reply client.go:22]: I think it's different from serverOptions

**Comment by StringNick** [comment]
added dial timeout, dial

**Comment by MrChang0** [comment]
emmmm...sorry for late comment. because of busy work.
1. with non-block connect, I think use  'unix.select' is too simple. unix.select is blocking api for wait fd result, with gev or other network library we just use 'Poller' for all fd including client fd. (btw go network is base non blocking api with select) the best way to manager client fd, we need manage fd state and use poller callback to replace 'unix.Select'.  like [muduo connector](https://github.com/chenshuo/muduo/blob/9aaeda812bde476d231495212e6cf331ae2792b4/muduo/net/Connector.cc#L158).
2. your code is not clear enough,  some duplicate code need clean, and emmmm... 
3. btw, I can see you know network, but not familiar with network library. I think you need see more network library. like muduo, libuv, libev...you will learn more from them.
4. beat wishes you can finish this pr.

**Comment by StringNick** [comment]
> emmmm...sorry for late comment. because of busy work.
> 
> 1. with non-block connect, I think use  'unix.select' is too simple. unix.select is blocking api for wait fd result, with gev or other network library we just use 'Poller' for all fd including client fd. (btw go network is base non blocking api with select) the best way to manager client fd, we need manage fd state and use poller callback to replace 'unix.Select'.  like [muduo connector](https://github.com/chenshuo/muduo/blob/9aaeda812bde476d231495212e6cf331ae2792b4/muduo/net/Connector.cc#L158).
> 2. your code is not clear enough,  some duplicate code need clean, and emmmm...
> 3. btw, I can see you know network, but not familiar with network library. I think you need see more network library. like muduo, libuv, libev...you will learn more from them.
> 4. beat wishes you can finish this pr.

yea this my first experience with it, i watched c libraries, i will try it on weekends to rewrite on poller, but problem in that i should use both kqueue and epoll, i will install vagrant and try it

**Comment by MrChang0** [comment]
> > emmmm...sorry for late comment. because of busy work.
> > 
> > 1. with non-block connect, I think use  'unix.select' is too simple. unix.select is blocking api for wait fd result, with gev or other network library we just use 'Poller' for all fd including client fd. (btw go network is base non blocking api with select) the best way to manager client fd, we need manage fd state and use poller callback to replace 'unix.Select'.  like [muduo connector](https://github.com/chenshuo/muduo/blob/9aaeda812bde476d231495212e6cf331ae2792b4/muduo/net/Connector.cc#L158).
> > 2. your code is not clear enough,  some duplicate code need clean, and emmmm...
> > 3. btw, I can see you know network, but not familiar with network library. I think you need see more network library. like muduo, libuv, libev...you will learn more from them.
> > 4. beat wishes you can finish this pr.
> 
> yea this my first experience with it, i watched c libraries, i will try it on weekends to rewrite on poller, but problem in that i should use both kqueue and epoll, i will install vagrant and try it

'i should use both kqueue and epoll' you needn't change Poller, juse reuse it. ex client connector with three state(simple) -'connecting, connected, disconnected', you can manage state in client 'onMessage'. when client is connecting, it should change to connected; when it is connected, it should callback user onMessage.......

**Comment by MrChang0** [comment]
> > > emmmm...sorry for late comment. because of busy work.
> > > 
> > > 1. with non-block connect, I think use  'unix.select' is too simple. unix.select is blocking api for wait fd result, with gev or other network library we just use 'Poller' for all fd including client fd. (btw go network is base non blocking api with select) the best way to manager client fd, we need manage fd state and use poller callback to replace 'unix.Select'.  like [muduo connector](https://github.com/chenshuo/muduo/blob/9aaeda812bde476d231495212e6cf331ae2792b4/muduo/net/Connector.cc#L158).
> > > 2. your code is not clear enough,  some duplicate code need clean, and emmmm...
> > > 3. btw, I can see you know network, but not familiar with network library. I think you need see more network library. like muduo, libuv, libev...you will learn more from them.
> > > 4. beat wishes you can finish this pr.
> > 
> > 
> > yea this my first experience with it, i watched c libraries, i will try it on weekends to rewrite on poller, but problem in that i should use both kqueue and epoll, i will install vagrant and try it
> 
> 'i should use both kqueue and epoll' you needn't change Poller, juse reuse it. ex client connector with three state(simple) -'connecting, connected, disconnected', you can manage state in client 'onMessage'. when client is connecting, it should change to connected; when it is connected, it should callback user onMessage.......

I see sth worry, client connected state need write handler (EventWrite) and not onMessage(EventRead). QWQ....

**Comment by StringNick** [comment]
в7o6i

> > > > emmmm...sorry for late comment. because of busy work.
> > > > 
> > > > 1. with non-block connect, I think use  'unix.select' is too simple. unix.select is blocking api for wait fd result, with gev or other network library we just use 'Poller' for all fd including client fd. (btw go network is base non blocking api with select) the best way to manager client fd, we need manage fd state and use poller callback to replace 'unix.Select'.  like [muduo connector](https://github.com/chenshuo/muduo/blob/9aaeda812bde476d231495212e6cf331ae2792b4/muduo/net/Connector.cc#L158).
> > > > 2. your code is not clear enough,  some duplicate code need clean, and emmmm...
> > > > 3. btw, I can see you know network, but not familiar with network library. I think you need see more network library. like muduo, libuv, libev...you will learn more from them.
> > > > 4. beat wishes you can finish this pr.
> > > 
> > > 
> > > yea this my first experience with it, i watched c libraries, i will try it on weekends to rewrite on poller, but problem in that i should use both kqueue and epoll, i will install vagrant and try it
> > 
> > 
> > 'i should use both kqueue and epoll' you needn't change Poller, juse reuse it. ex client connector with three state(simple) -'connecting, connected, disconnected', you can manage state in client 'onMessage'. when client is connecting, it should change to connected; when it is connected, it should callback user onMessage.......
> 
> I see sth worry, client connected state need write handler (EventWrite) and not onMessage(EventRead). QWQ....

i think i found way to solve it with poller, i will try push it today

**Comment by MrChang0** [comment]
> в7o6i
> 
> > > > > emmmm...sorry for late comment. because of busy work.
> > > > > 
> > > > > 1. with non-block connect, I think use  'unix.select' is too simple. unix.select is blocking api for wait fd result, with gev or other network library we just use 'Poller' for all fd including client fd. (btw go network is base non blocking api with select) the best way to manager client fd, we need manage fd state and use poller callback to replace 'unix.Select'.  like [muduo connector](https://github.com/chenshuo/muduo/blob/9aaeda812bde476d231495212e6cf331ae2792b4/muduo/net/Connector.cc#L158).
> > > > > 2. your code is not clear enough,  some duplicate code need clean, and emmmm...
> > > > > 3. btw, I can see you know network, but not familiar with network library. I think you need see more network library. like muduo, libuv, libev...you will learn more from them.
> > > > > 4. beat wishes you can finish this pr.
> > > > 
> > > > 
> > > > yea this my first experience with it, i watched c libraries, i will try it on weekends to rewrite on poller, but problem in that i should use both kqueue and epoll, i will install vagrant and try it
> > > 
> > > 
> > > 'i should use both kqueue and epoll' you needn't change Poller, juse reuse it. ex client connector with three state(simple) -'connecting, connected, disconnected', you can manage state in client 'onMessage'. when client is connecting, it should change to connected; when it is connected, it should callback user onMessage.......
> > 
> > 
> > I see sth worry, client connected state need write handler (EventWrite) and not onMessage(EventRead). QWQ....
> 
> i think i found way to solve it with poller, i will try push it today, without changing Connection/EventLoop and etc

 btw you can read [user guide](https://github.com/uber-go/guide). when you finish your code, review it clam and carefully. you are first reviwer of your code.

**Comment by StringNick** [comment]
reworked on eventloop connector, waiting comments ty

File: connector/connection.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+	timeout time.Duration,
+	protocol connection.Protocol,
+	tw *timingwheel.TimingWheel,
+	idleTime time.Duration,
+	callBack connection.CallBack) (*Connection, error) {
+
+	fd, err := unixOpenConnect(network, address)
+	if err != nil {
+		if fd > 0 {
+			_ = unix.Close(fd)
```
Comment (Allenxuxu) [inline connector/connection.go:57]: I think `close(fd)` should be handle in function `unixOpenConnect `.
If `unixOpenConnect ` return error,  `fd` need be 0 （Close it inside the function `unixOpenConnect `）

**Comment by StringNick** [comment]
I need help, when i connecting to port is not exist, error is uint32 129/130 Poller.Event, how can I decode it to normal, or u can help me with that, I saw u got ready errors codes converted to ur interface

**Comment by Allenxuxu** [comment]
> I need help, when i connecting to port is not exist, error is uint32 129/130 Poller.Event, how can I decode it to normal, or u can help me with that, I saw u got ready errors codes converted to ur interface

Write unit test to show this.

**Comment by StringNick** [comment]
> > I need help, when i connecting to port is not exist, error is uint32 129/130 Poller.Event, how can I decode it to normal, or u can help me with that, I saw u got ready errors codes converted to ur interface
> 
> Write unit test to show this.

I pushed test, when not existing listner, it returns 129, sometime 130

**Comment by Allenxuxu** [comment]
image

129 == 0X81 == (EventErr | EventRead)
130 == 0X82 == (EventErr | EventWrite)

image

This approach is not correct.

You should be like this.
https://github.com/Allenxuxu/gev/blob/631f7252a9ca88315d327ed4a26d0f97947abdad/connection/connection.go#L167

Looking forward to your optimization.

**Comment by StringNick** [comment]
> image
> 
> 129 == 0X81 == (EventErr | EventRead)
> 130 == 0X82 == (EventErr | EventWrite)
> 
> image
> 
> This approach is not correct.
> 
> You should be like this.
> 
> https://github.com/Allenxuxu/gev/blob/631f7252a9ca88315d327ed4a26d0f97947abdad/connection/connection.go#L167
> 
> Looking forward to your optimization.

done

File: connector/connection.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+				c.result <- err
+				return
+			}
+
+			c.Connection = connection.New(c.fd, c.loop, sa, c.protocol, c.tw, c.idleTime, c.callBack)
+			c.state = connectedConnectionSocketState
+			c.result <- nil
+			c.Connection.HandleEvent(fd, events)
+			return
+		} else if events&poller.EventWrite&poller.EventErr == 0 || events&poller.EventRead&poller.EventErr == 0 {
```
Comment (Allenxuxu) [inline connector/connection.go:151]: This is not a good way to deal with it. You should handle the eventErr first.

**Review by Allenxuxu — CHANGES_REQUESTED** [review]

File: connector/connection.go (hunk 7 above)
Comment (Allenxuxu) [reply connector/connection.go:151]: And please ensure that all unit tests pass.
https://github.com/Allenxuxu/gev/pull/83/checks?check_run_id=2759493948#step:6:255

**Comment by StringNick** [comment]
waiting approval for checking by linter

**Comment by StringNick** [comment]
I can't saw error of failing unit-testing

**Comment by Allenxuxu** [comment]
> I can't saw error of failing unit-testing

https://github.com/StringNick/gev/actions 
Try opening Action in your repo. And you should pass all unit test in your PC before you push.

**Comment by Allenxuxu** [comment]
https://github.com/Allenxuxu/gev/tree/connector

I created a new branch，You can close this PR and create a new PR to branch `connector`, when the feature is stable, we will merge it into the master.
We(@StringNick @MrChang0 @Allenxuxu ) can work more convenient in branch `connector`.

**Pull request closed without merging**